server/src/main/java/org/opensearch/gateway/TransportNodesBulkListGatewayStartedShards.java

```diff
@@ -0,0 +1,429 @@
+                ShardStateMetadata shardStateMetadata = ShardStateMetadata.FORMAT.loadLatestState(
+                    logger,
+                    namedXContentRegistry,
+                    nodeEnv.availableShardPaths(shardId)
+                );
+                if (shardStateMetadata != null) {
+                    if (indicesService.getShardOrNull(shardId) == null) {
+                        final String customDataPath = TransportNodesGatewayStartedShardHelper.getCustomDataPathForShard(
+                            logger,
+                            shardId,
+                            shardToCustomDataPathEntry.getValue(),
+                            settings,
+                            clusterService
+                        );
+                        // we don't have an open shard on the store, validate the files on disk are openable
+                        Exception shardCorruptionException = TransportNodesGatewayStartedShardHelper.getShardCorruption(
+                            logger,
+                            nodeEnv,
+                            shardId,
+                            shardStateMetadata,
+                            customDataPath
+                        );
+                        if (shardCorruptionException != null) {
+                            String allocationId = shardStateMetadata.allocationId != null ? shardStateMetadata.allocationId.getId() : null;
+                            shardsOnNode.put(
+                                shardId,
+                                new NodeGatewayStartedShards(allocationId, shardStateMetadata.primary, null, shardCorruptionException)
+                            );
+                            continue;
+                        }
+                    }
+                    logger.debug("{} shard state info found: [{}]", shardId, shardStateMetadata);
+                    String allocationId = shardStateMetadata.allocationId != null ? shardStateMetadata.allocationId.getId() : null;
+                    final IndexShard shard = indicesService.getShardOrNull(shardId);
+                    shardsOnNode.put(
+                        shardId,
+                        new NodeGatewayStartedShards(
+                            allocationId,
+                            shardStateMetadata.primary,
+                            shard != null ? shard.getLatestReplicationCheckpoint() : null
+                        )
+                    );
+                } else {
+                    logger.trace("{} no local shard info found", shardId);
+                    shardsOnNode.put(shardId, new NodeGatewayStartedShards(null, false, null));
+                }
+            } catch (Exception e) {
+                throw new OpenSearchException("failed to load started shards", e);
+            }
+        }
+        return new BulkOfNodeGatewayStartedShards(clusterService.localNode(), shardsOnNode);
+    }
+
+    /**
+     * The nodes request.
```

Comment:
Can you just re-check if we need node**S** classes ?

Reply:
I think this is needed to satisfy the types required by the base class. I could be wrong due to my limited understanding of java and the code base.